### PR TITLE
Implemented type attribute for OscillatorNode interface

### PR DESCRIPTION
**Pull request by PurpleHairEngineer**
Please see https://github.com/servo/media/pull/242 for implementation of the new oscillator node message.

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23282)

### CODE REVIEW

**Review by jdm — CHANGES_REQUESTED** [review]
The implementation changes appear to be missing.

**Comment by PurpleHairEngineer** [comment]
@jdm The implementation changes are in https://github.com/servo/media/pull/242. What do you mean when you say they appear to be missing? I didn't know how to submit a pull request for both servo/servo and servo/media at the same time, which is why I split it up into two different pull requests.

**Comment by jdm** [comment]
This change does not compile right now, since there are missing SetType and GetType methods for OscillatorNode.

**Comment by PurpleHairEngineer** [comment]
@jdm My bad. I think I added the correct methods, although GetType was called Type in biquadfilternode.rs, so I kept the same style and just called it Type.

File: components/script/dom/webidls/OscillatorNode.webidl
Comment (ferjm) [inline components/script/dom/webidls/OscillatorNode.webidl]: We need to uncomment this line as well, so we can throw an `InvalidStateError` exception when trying to set the type to `OscillatorType::Custom`, as [specified](https://webaudio.github.io/web-audio-api/#dom-oscillatornode-type)

**Review by ferjm — CHANGES_REQUESTED** [review]
Thanks for the PR :)

There are still a few [build errors](https://travis-ci.com/servo/servo/jobs/196105948#L2445) that need to be fixed.

**Comment by PurpleHairEngineer** [comment]
@ferjm I fixed a few of the build errors, but I can't figure out why it can't find the get and set methods: error[E0599]: no method named `get` found for type `dom::bindings::codegen::Bindings::OscillatorNodeBinding::OscillatorType` in the current scope. This is on line 120 of oscillatornode.rs now (and line 125 for the set method).

**Comment by ferjm** [comment]
`dom::bindings::codegen::Bindings::OscillatorNodeBinding::OscillatorType` is an enum and it has no `get` or `set` methods.

You likely got these method calls from the `BiquadFilterNode.filter` related code that you were taking as example. If you look closer into that code, you could see that the we are using a [Cell](https://doc.rust-lang.org/std/cell/struct.Cell.html) to provide [interior mutability](https://doc.rust-lang.org/nightly/reference/interior-mutability.html) because the setters and getters of DOM struct members are getting a reference (`&self`) and not a mutable reference (`&mut self`). `Cell`s expose the [get](https://doc.rust-lang.org/std/cell/struct.Cell.html#method.get) and [set](https://doc.rust-lang.org/std/cell/struct.Cell.html#method.get) methods you were calling.

We need exactly the same thing for `OscillatorNode.oscillator_type`. This is, to wrap it in a `Cell` to get interior mutability.

**Comment by PurpleHairEngineer** [comment]
@ferjm There are build errors for this, but a main reason seems to be that the OscillatorNodeMessage hasn't been implemented yet (see https://github.com/servo/media/pull/242). Is there anything else I can do for this pull request before that one is merged?

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]: No need for this temporary variable.

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]: You can directly do:

```rust
oscillator_type: Cell::new(options.type_)
```

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]: Because `type` has the `SetterThrows` attribute on the WebIDL, the generated setter returns an `ErrorResult`. That's the reason why you are getting [this](https://travis-ci.com/servo/servo/jobs/196337040#L2529) build error. The function type should be:

```rust
fn SetType(&self, type_: OscillatorType) -> ErrorResult
```

Also, one nit: let's use `type_` instead of `oscillator` for the `OscillatorType` argument, please.

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]: According to the [spec](https://webaudio.github.io/web-audio-api/#dom-oscillatornode-type) we need to throw an `InvalidStateError` when trying to set a custom type.

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]: `OscillatorNode` has no `node` member. In this case, we are inheriting from `AudioScheduledSourceNode` (which inherits from `AudioNode`). So what we have is a `source_node` member, which has a `node()` getter. This should be:

```rust
self.source_node.node().message(...)
```

**Review by ferjm — CHANGES_REQUESTED** [review]
servo/media#242 was just merged. You should be able to update servo-media here to get the latest changes by running `./mach cargo-update -p servo-media`.

**Comment by PurpleHairEngineer** [comment]
@ferjm Could you please take a look at this? I am getting build errors, but they all seem to be for files that are not modified for this pull request.

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]:
```
error[E0432]: unresolved import `servo_media::audio::oscillator_node::OscillatorNodeMessage`
  --> components/script/dom/oscillatornode.rs:22:5
   |
22 | use servo_media::audio::oscillator_node::OscillatorNodeMessage;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^---------------------
   |     |                                    |
   |     |                                    help: a similar name exists in the module: `OscillatorNode`
   |     no `OscillatorNodeMessage` in `audio::oscillator_node`
```

You still need to update `servo-media` to get the changes where you introduced `OscillatorNodeMessage`.

Run:
```
./mach cargo-update -p servo-media
```

And commit the changes done to the `Cargo.lock` file.

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]:
```
error[E0308]: mismatched types
   --> components/script/dom/oscillatornode.rs:124:49
    |
124 |     fn SetType(&self, type_: OscillatorType) -> ErrorResult {
    |        -------                                  ^^^^^^^^^^^ expected enum `std::result::Result`, found ()
    |        |
    |        this function's body doesn't return
    |
    = note: expected type `std::result::Result<(), dom::bindings::error::Error>`
               found type `()`
```

This function expects an `ErrorResult` and at this point everthing should be fine, so we need to return `Ok(())` here.

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]:
```
error[E0308]: mismatched types
   --> components/script/dom/oscillatornode.rs:126:21
    |
126 |         if type_ == "custom" {
    |                     ^^^^^^^^ expected enum `dom::bindings::codegen::Bindings::OscillatorNodeBinding::OscillatorType`, found reference
    |
    = note: expected type `dom::bindings::codegen::Bindings::OscillatorNodeBinding::OscillatorType`
               found type `&'static str`
```

WebIDL defined enums generate Rust enums with each entry's first letter capitalized. We need to compare with `OscillatorType::Custom` in this case.

File: components/script/dom/oscillatornode.rs
Comment (ferjm) [inline components/script/dom/oscillatornode.rs]:
```
error[E0599]: no variant or associated item named `OscillatorNode` found for type `servo_media_audio::node::AudioNodeMessage` in the current scope
   --> components/script/dom/oscillatornode.rs:131:40
    |
131 |             .message(AudioNodeMessage::OscillatorNode(
    |                                        ^^^^^^^^^^^^^^ variant or associated item not found in `servo_media_audio::node::AudioNodeMessage`
```

Updating `servo-media` will get the `AudioNodeMessage::OscillatorNode` variant as well.

**Review by ferjm — CHANGES_REQUESTED** [review]

File: components/script/dom/oscillatornode.rs
Comment (PurpleHairEngineer) [reply components/script/dom/oscillatornode.rs]: Woops, I had done that locally but didn't know I needed to commit the changes to Cargo.lock. Now that I've done that, it says this branch has merge conflicts that need to be resolved.

**Comment by PurpleHairEngineer** [comment]
@ferjm Now that I've done committed the changes to Cargo.lock, it says this branch has merge conflicts that need to be resolved. I'm not sure how to resolve them because I'm not sure what the correct version of Cargo.lock is?

**Comment by jdm** [comment]
@PurpleHairEngineer I recommend performing a rebase, and when it encounters the merge conflict with Cargo.lock using `git rebase --skip`, and once the rebase is complete then run the cargo update command that @ferjm mentioned. This will mean that the Cargo.lock will be in a pristine state, and then the update will be applied on top of it.

**Comment by PurpleHairEngineer** [comment]
@jdm I ran `git rebase -i origin/master` and closed the text editor without changing everything. I got the message `Successfully rebased and updated refs/heads/OscillatorNodeType.` It did not encounter the merge conflict. Did I enter the wrong command?

**Comment by ferjm** [comment]
@PurpleHairEngineer did you fetch upstream before rebasing?

```
git fetch upstream
git rebase upstream/master
(once the Cargo.lock conflict appears) git rebase --skip
```

These commands assume that you have a [remote](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes) named `upstream`. To check if you have an `upstream` remote, run `git remote -v`, you should see something like `upstream git@github.com:servo/servo.git (fetch)`. If you don't have it, you can run `git remote add upstream git@github.com:servo/servo.git`. And then run the fetch and rebase commands above.

In any case, I recommend waiting for #23294 to be merged, as updating servo-media before that one will throw additional build errors.

**Comment by ferjm** [comment]
@PurpleHairEngineer #23294 was already merged, so this PR should be ready for merge after rebasing. Let me know if you need any help with the rebase or if you want me to do it? Thanks.

**Comment by ferjm** [comment]
I went ahead and rebase and squashed the commits

**Comment by CYBAI** [comment]
```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "OscillatorNode interface: attribute type", 
    "test": "/webaudio/idlharness.https.window.html", 
    "line": 160167, 
    "action": "test_result", 
    "expected": "FAIL"
}
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "OscillatorNode interface: new OscillatorNode(context) must inherit property \"type\" with the proper type", 
    "test": "/webaudio/idlharness.https.window.html", 
    "line": 160173, 
    "action": "test_result", 
    "expected": "FAIL"
}
```

**Pull request merged**
